### ISSUE TITLE
[GEP-20] Reduce Cross-zonal Traffic via Topology Aware Hints

### ISSUE THREAD

**Issue by timuthy** [gardener/gardener#6718]
**How to categorize this issue?**

/area high-availability
/area networking
/area cost
/kind enhancement

**What would you like to be added**:
Gardener should provide means to reduce cross-zonal network traffic within HA control-planes (with `zone` tolerance). For now, we plan to stick to standard Kubernetes services and features and not introduce any service meshes that bring advanced routing capabilities.
A Kubernetes in-built feature to fulfill this requirement is called [Topology Aware Hints](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/).
All relevant services must be annotated with `service.kubernetes.io/topology-aware-hints: auto` to enable the hints:

Seed cluster: https://github.com/gardener/gardener/pull/7191
- [x] etcd-{main,events}-client 
   - [x] Allow specifying etcd client Service annotations: https://github.com/gardener/etcd-druid/issues/435
   - [x] Allow specifying etcd client Service labels: https://github.com/gardener/etcd-druid/pull/485
- [x] kube-apiserver
- [x] gardener-resource-manager
- [x] vpa-webhook

Seed cluster (extensions)
- [x] csi-snapshot-validation
   - [x] gardener-extension-provider-aws: https://github.com/gardener/gardener-extension-provider-aws/pull/703
   - [x] gardener-extension-provider-gcp: https://github.com/gardener/gardener-extension-provider-gcp/pull/574
   - [x] gardener-extension-provider-azure: https://github.com/gardener/gardener-extension-provider-azure/pull/649
   - [x] gardener-extension-provider-openstack: https://github.com/gardener/gardener-extension-provider-openstack/pull/589
   - [x] gardener-extension-provider-alicloud: https://github.com/gardener/gardener-extension-provider-alicloud/pull/587
   - [x] gardener-extension-provider-vsphere: https://github.com/gardener/gardener-extension-provider-vsphere/pull/391
- [x] gardener-extension-provider-aws: https://github.com/gardener/gardener-extension-provider-aws/pull/703
  - Details : provider-aws runs a Shoot webhook that mutates the `addons-nginx-ingress-controller` ConfigMap.
- [x] gardener-extension-provider-alicloud: https://github.com/gardener/gardener-extension-provider-alicloud/pull/587
  - Details: provider-alicloud runs a Shoot webhook that mutates the `vpn-shoot` and `addons-nginx-ingress-controller` Services and the `metrics-server` Deployment.
- [x] gardener-extension-provider-vsphere: https://github.com/gardener/gardener-extension-provider-vsphere/pull/391
  - Details: provider-vsphere runs a Shoot webhok that mutates the `metrics-server` Deployment.
- [x] gardener-extension-networking-cilium: https://github.com/gardener/gardener-extension-networking-cilium/pull/172
  - Details: networking-cilium runs a Shoot webhook that mutates the `node-local-dns` DaemonSet and ConfigMap.

Garden runtime cluster: https://github.com/gardener/gardener/pull/7729
- [x] virtual-garden-etcd-{main,events}-client
- [x] virtual-garden-kube-apiserver
- [x] gardener-apiserver
- [x] gardener-admission-controller

Garden runtime cluster (extensions):
- Extension admission components
  - [x] gardener-extension-admission-alicloud: https://github.com/gardener/gardener-extension-provider-alicloud/pull/591
  - [x] gardener-extension-admission-aws: https://github.com/gardener/gardener-extension-provider-aws/pull/732
  - [x] gardener-extension-admission-azure: https://github.com/gardener/gardener-extension-provider-azure/pull/657
  - [x] gardener-extension-admission-calico: https://github.com/gardener/gardener-extension-networking-calico/pull/258
  - [x] gardener-extension-admission-cilium: https://github.com/gardener/gardener-extension-networking-cilium/pull/176
  - [x] gardener-extension-admission-gcp: https://github.com/gardener/gardener-extension-provider-gcp/pull/582
  - [x] gardener-extension-admission-openstack: https://github.com/gardener/gardener-extension-provider-openstack/pull/599
  - [x] gardener-extension-admission-shoot-dns-service: https://github.com/gardener/gardener-extension-shoot-dns-service/pull/203
  - [x] gardener-extension-shoot-auditlog-admission
  - [x] gardener-extension-validator-vsphere: https://github.com/gardener/gardener-extension-provider-vsphere/pull/397

- In K8s 1.27 the `service.kubernetes.io/topology-aware-hints=auto` annotation is deprecated in favor of `service.kubernetes.io/topology-mode=auto` (ref https://github.com/kubernetes/kubernetes/pull/116522). Usages of the deprecated annotation in gardener/gardener and extensions have to be adapted.
  - [x] gardenlet and gardener-operator: https://github.com/gardener/gardener/pull/7933
  - [x] gardener-extension-provider-alicloud: https://github.com/gardener/gardener-extension-provider-alicloud/pull/601
  - [x] gardener-extension-provider-aws: https://github.com/gardener/gardener-extension-provider-aws/pull/750
  - [x] gardener-extension-provider-azure: https://github.com/gardener/gardener-extension-provider-azure/pull/686
  - [x] gardener-extension-provider-gcp: https://github.com/gardener/gardener-extension-provider-gcp/pull/597
  - [x] gardener-extension-provider-openstack: https://github.com/gardener/gardener-extension-provider-openstack/pull/625
  - [x] gardener-extension-provider-vsphere: https://github.com/gardener/gardener-extension-provider-vsphere/pull/413
  - [x] gardener-extension-networking-calico: https://github.com/gardener/gardener-extension-networking-calico/pull/271
  - [x] gardener-extension-networking-cilium: https://github.com/gardener/gardener-extension-networking-cilium/pull/188
  - [x] gardener-extension-shoot-auditlog-service
  - [x] gardener-extension-shoot-dns-service: https://github.com/gardener/gardener-extension-shoot-dns-service/pull/211
  
Cilium:
- [x] Add support for topology-aware traffic routing in ManagedSeeds using cilium (ManagedSeeds without kube-proxy): https://github.com/gardener/gardener-extension-networking-cilium/pull/185

Monitoring:
- [ ] Add monitoring for the cross-zone network traffic. This is required to confirm that with the newly introduced changes we reduce the cross-zone traffic. (ref https://github.tools.sap/kubernetes/networking-backlog/issues/113)

Follow-up items:
- [x] Investigate if kube-apiserver Service is really eligible/safe for topology-aware routing (see https://github.com/gardener/gardener/issues/6718#issuecomment-1436483874)
  - The external traffic is already randomized on istio-ingressgateway level according to the cloud provider load balancer's algorithm. The question is whether the internal traffic to the kube-apiserver Service can cause such imbalance.
  - No issues were observed.
- [x] Follow-up in the upstream discussions to have the deterministic topology-aware routing implemented in Kubernetes. This will allow us to drop the `endpoint-slice-hints.resources.gardener.cloud` webhook. (ref https://github.com/kubernetes/kubernetes/issues/113731)
  - This is an ongoing item. The discussion is tracked in the above-mentioned issue.

**Why is this needed**:
Creating HA control-planes distributed over different availability zones will by default result in an unpredictable amount of cross-zonal network traffic (e.g. `kube-apiserver` in **Zone-A** reaches out to `etcd` member in **Zone-B**). This usually increases latency and produces unwanted costs because Hyperscalers charge this type of traffic additionally.

**Comment by ialidzhikov** [comment]
In the last days I had a look into the [TopologyAwareHints](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/). I will try to summarize my understanding in this comment. Most of the points raised in this comment are also raised by @unmarshall in https://github.com/kubernetes/kubernetes/issues/110714 (the issue title is specific but discussion in the issue itself is more comprehensive).

**What `TopologyAwareHints` offers?**

The EndpointSlice controller of kube-controller-manager is responsible to setting the `hints` on the EndpointSlice resources. The used algorithm tries to achieve a balanced distribution of endpoints among the zones:

> The controller allocates a proportional amount of endpoints to each zone. This proportion is based on the allocatable CPU cores for nodes running in that zone. For example, if one zone had 2 CPU cores and another zone only had 1 CPU core, the controller would allocate twice as many endpoints to the zone with 2 CPU cores.

And the following safeguard mechanism is implemented when it is not possible to achieve a balanced distribution:
> **Impossible to achieve balanced allocation**: In some cases, it will be impossible to achieve a balanced allocation of endpoints among zones. For example, if zone-a is twice as large as zone-b, but there are only 2 endpoints, an endpoint allocated to zone-a may receive twice as much traffic as zone-b. The controller does not assign hints if it can't get this "expected overload" value below an acceptable threshold for each zone. Importantly this is not based on real-time feedback. It is still possible for individual endpoints to become overloaded.

The used algorithm is based on the following heuristic: it assumes that the estimated traffic for a Service from a zone is proportional to the zone's size (the allocatable CPU of Nodes in that zone). When it cannot achieve a balanced allocation that is proportional to the zones' CPU ratio within the allowed [overload threshold](https://github.com/kubernetes/kubernetes/blob/c3259df151118f7f0d05d9d4c773508582a3769a/pkg/controller/endpointslice/topologycache/topologycache.go#L31-L33), it does not add `hints` or removes already existing `hints`. This is implemented as safeguard mechanism to prevent overloading certain endpoints.

**In our setup**

- The clients and the servers are well-known and in most of the cases the traffic a component receives does not depend on the zone's allocatable CPU.

   In our setup the clients and the servers are well-known. For example the traffic between the following components is a candidate to be topology aware:
   - `etcd-main`/`etcd/events` and `kube-apiserver`
   - `kube-apiserver` and all the webhook servers running in the Seed (`gardener-resource-manager`, `vpa-admission-controller`, `csi-snapshot-validation`, extensions with a Shoot webhook)
   - `kube-apiserver` and all the components that interact with it (`istio-ingressgateway`, `gardenlet`, etc.)
   - etc.

   Let's take the communication between `kube-apiserver` and `etcd-main`. `etcd-main` runs with 3 replicas spread on the 3 zones, kube-apiserver runs with at least 3 replicas spread on the zones. We want to enforce the traffic between the 2 components to be topology aware - traffic from kube-apiserver(s) from zone-a to be routed only to the `etcd-main` replica in zone-a. With `TopologyAwareHints` whether the etcd-main Service will be topology-aware or not depends on the zones' allocatable CPU. In our setup we rather want this operation to be deterministic and the `etcd-main` Service to be always topology aware. We also leverage VPA for most of the above-mentioned components. As a safeguard mechanism, in case a replica is overloaded, VPA should take care to autoscale.

- We use cluster-autoscaler to autoscale to Node's count and we don't have a way to guarantee that the zones' allocatable CPU will be within the given threshold.

   We run different worker pools with different machine types (different allocatable CPU) and we use the cluster-autoscaler. 

   An example of real world multi-zone Seed:
   ```
   $ k get nodes -o custom-columns='NAME:.metadata.name,ZONE:.metadata.labels.topology\.kubernetes\.io\/zone,CPU:.status.allocatable.cpu,GROUP:.metadata.labels.worker\.gardener\.cloud\/pool'
   NAME                                          ZONE         CPU      GROUP
   <omitted>                                     eu-west-1a   15920m   cpu-worker
   <omitted>                                     eu-west-1a   15920m   cpu-worker
   <omitted>                                     eu-west-1a   7920m    etcd-compaction
   <omitted>                                     eu-west-1b   7920m    cpu-worker-etcd
   <omitted>                                     eu-west-1b   15920m   cpu-worker
   <omitted>                                     eu-west-1a   7920m    cpu-worker-etcd
   <omitted>                                     eu-west-1c   15920m   cpu-worker
   <omitted>                                     eu-west-1c   15920m   cpu-worker
   <omitted>                                     eu-west-1c   7920m    cpu-worker-etcd
   
   Total CPU = 111280
   
   CPU per zone:
   eu-west-1a = 47680m
   eu-west-1b = 23840m
   eu-west-1c = 39760m
   
   CPU ratios
   eu-west-1a = 0.42
   eu-west-1b = 0.21
   eu-west-1c = 0.35
   ```

   For the above cluster, if we want to enable TopologyAwareHints for a Service with 3 endpoints:
   => zone eu-west-1a is 2x bigger than eu-west-1b, kube-controller-manager has to ensure that eu-west-1a gets 2x more endpoints than eu-west-1b => impossible to achieve balanced allocation => `hints` are not set

   According to my understanding, for a Service with 3 endpoints on 3 zone cluster, if the allotable CPU on a zone is more than 20% bigger than the allocatable CPU on other zone, then `hints` are not set for the given Service EndpointSlices.

TL;DR I am not sure if `TopologyAwareHints` is what we are looking for (in its current feature offering state) and whether it will lead to any noticeable improvements if we proceed with it.

cc @vlerenc @unmarshall @timuthy @rfranzke @ScheererJ @DockToFuture 

---

Docs: https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2433-topology-aware-hints/README.md

**Comment by vlerenc** [comment]
Is it possible to set/mutate those hints ourself, so that we get what we need, because the logic to not set the hints is indeed flawed in our case (unless there is no endpoint in that zone that is, but it shouldn't depend on the CPU capacity per zone - that's pointless/wrong in our case).

**Comment by DockToFuture** [comment]
FYI, we run cilium shoots without `kube-proxy` and we also wanted to remove `kube-proxy` from calico shoots at a later point in time.

**Comment by unmarshall** [comment]
>Is it possible to set/mutate those hints ourself, so that we get what we need, because the logic to not set the hints is indeed flawed in our case

AFAIK even if you set this explicitly, [reconciler](https://github.com/kubernetes/kubernetes/blob/29b856b658337ac94cf808d7ce3925b0af527f86/pkg/controller/endpointslice/reconciler.go#L264-L270) will again change the hints based on the values computed in [topologycache](https://github.com/kubernetes/kubernetes/blob/29b856b658337ac94cf808d7ce3925b0af527f86/pkg/controller/endpointslice/topologycache/topologycache.go). So most probably it will again be overwritten. However we can test this out and also confirm it.

>FYI, we run cilium shoots without kube-proxy and we also wanted to remove kube-proxy from calico shoots at a later point in time.

Cilium has support for TSC (https://github.com/cilium/cilium/pull/17929)

**Comment by ialidzhikov** [comment]
We had a sync last week on this topic. The "consensus" was to ask the maintainers feedback for our use case and try to contribute it. There was also the suggestion to evaluate a webhook approach that on every update ensures the EndpointSlice hints.

~~I also noticed the [`ServiceTopology`](https://kubernetes.io/docs/concepts/services-networking/service-topology/) - it was deprecated in favor of TopologyAwareHints. I am mentioning it because it seem to have the feature we are looking for - see https://kubernetes.io/docs/concepts/services-networking/service-topology/#using-service-topology. We could evaluate it if the missing feature in TopologyAwareHints cannot be added in reasonable amount of time.~~
_Update: Actually we cannot use the `ServiceTopology` feature as the corresponding fields are removed in 1.22 with https://github.com/kubernetes/kubernetes/pull/102412._

**Comment by vlerenc** [comment]
We may need that feature not only because of **latency** and **cost**, but also for **functional/HA** reasons. Example:

Zone 1: KAPI + KCM leader talks to KAPI in zone 3 (random, doesn't matter)
Zone 2: KAPI + KCM loser talks to KAPI in zone 1 (random, important here)
Zone 3: KAPI

Then we kill zone 1. KCM leader is dead. So far so good. But KCM loser tries to talk to KAPI in zone 1. That's suboptimal. First, it does so only after the lease timeout (expected), but then it will wait also for the TCP connection timeout (unexpected). Since we have a very long `nodeMonitorGracePeriod`, the endpoint of the KAPI in zone 1 isn't even removed and then it may even try again (unexpected). The result is a very long KCM downtime with repercussions for the shoot cluster workload as nothing moves forward there, too.

All K8s services with explicit master nodes don't have that kind of problem, because all of their components speak node-internally, but we have enlarged our failure domain this way. With topology routing hints, this wouldn't have happened.

**Comment by ialidzhikov** [comment]
Last week when I was documenting prerequisites for making a Service topology-aware, I recalled:
> - This may not work well with autoscaling. For example, if a lot of traffic is originating from a single zone, only the endpoints allocated to that zone will be handling that traffic. That could result in [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) either not picking up on this event, or newly added pods starting in a different zone.

The above constraint is taken from https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/. The current list of constraints that I collected - ref https://github.com/gardener/gardener/pull/7191/files#diff-cd3d2b2108694a722dfe844731317f641ed644535882192558cbf22099c2601aR51-R54.

Currently the kube-apiserver is scaled by HVPA (when the corresponding feature gates are enabled) which is HPA + VPA AFAIK. So, I think theoretically we could have similar issues for the kube-apiserver, where a replica in a zone is overloaded, HPA kicks in and scales an replica in another zone. The overloaded kube-apiserver replica remains overloaded.

If we follow the constraints strictly, we should rather not make the kube-apiserver Service topology-aware. WDYT?

**Comment by unmarshall** [comment]
>If we follow the constraints strictly, we should rather not make the kube-apiserver Service topology-aware. WDYT?

This is a documented limitation and also points to just one case where a load within one zone is proportionally much higher than the rest of the zones. If you disable TAH completely for KAPI then you end up adding latency and cost for all scenarios. IMO it is better to salvage latency and cost in general (exceptional cases will always be there).

**Comment by vlerenc** [comment]
Gut feeling is similar as @unmarshall 's. If we don't have it, we pay with latency and cost at all times and the functional penalty now on top in case of an outage.

The standard behaviour is to remove the hints in the case @ialidzhikov mentioned (as there then would be more endpoints in one zone, but all zones have the same amount of allocatable CPUs), so that's only in combination with our own web hook, right?

Maybe we need to switch it on and learn/see what happens, trigger the corner cases, see what we can do about them. Possibly we can do something in the web hook or elsewhere, but without venturing there and trying the feature out, we won't know.

**Comment by ialidzhikov** [comment]
/close
as all subtasks related to the core team are resolved

There is still 1 open subtask about the monitoring topic but it seems to be already in progress by the networking folks and has a dedicated issue for tracking.
